### ISSUE TITLE
Implement gap filling

### ISSUE THREAD

**Issue by wolffcm** [influxdata/influxdb_iox#6594]
This issue is intended to track implementation progress of gap-filling:
https://github.com/influxdata/idpe/issues/16763

Technical design here:
https://docs.google.com/document/d/1vIcs9uhlCX_AkD9bemcDx-YhBOVe_TW5sBbXtKCHIfk/edit

Since community discussion on whether to upstream into DataFusion seems divided, I'll be implementing this in `influxdb_iox`.
https://github.com/apache/arrow-datafusion/issues/4809

The steps will be roughly:
- [ ] Create the scalar UDF `DATE_BIN_GAPFILL`
- [ ] Create an extension logical plan node `GapFill`
- [ ] Implement the optimizer rule to create the `GapFill` node
- [ ] Implement physical execution node for gap filling
- [ ] Add `LOCF` scalar function and semantics to `GapFill`

**Comment by stuartcarnie** [comment]
@wolffcm will the `DATE_BIN_GAPFILL` function support the `ORIGIN` parameter, as described by the [`DATE_BIN` function](https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-BIN). 

It is implemented in DataFusion's implementation of `DATE_BIN` here:

https://github.com/apache/arrow-datafusion/blob/9a6b76e3506151ec8fa12a8adfe559b4dc767208/datafusion/physical-expr/src/datetime_expressions.rs#L304-L317

This parameter would be required for complete `GROUP BY TIME()` compatibility, as it represents the `offset_interval` parameter of the [`TIME` function](https://docs.influxdata.com/influxdb/v1.8/query_language/explore-data/#advanced-group-by-time-syntax) in InfluxQL.

**Comment by wolffcm** [comment]
@stuartcarnie 
> will the DATE_BIN_GAPFILL function support the ORIGIN parameter, as described by the [DATE_BIN function](https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-BIN).

Yes, it will be supported. The signature for `DATE_BIN_GAPFILL` is intended to be identical to `DATE_BIN`, so it should be a drop-in replacement for many queries.

At a high-level the planner rule just does this:
- updates an `Aggregate` node to change `DATE_BIN_GAPFILL` to `DATE_BIN`
- creates a new `GapFill` plan node that succeeds the aggregate to fill in the gaps.

This reminds me to add tests where the origin is something more interesting that the epoch.

**Comment by alamb** [comment]
I wonder if this ticket should be closed now as "done as much as we plan"? Or are there still items we plan to complete (like https://github.com/influxdata/influxdb_iox/issues/7528) before claiming success?

**Comment by alamb** [comment]
> With that said, I'd be fine with closing this issue and making https://github.com/influxdata/influxdb_iox/issues/7528 a stand-alone issue if it would reduce friction.

It seems like this would be a good idea to make the tickets better reflect the state of things . What do you think @gavincabbage ?

**Comment by gavincabbage** [comment]
Thanks for calling this out. I agree we should tidy things up here. I will close this epic and refile the remaining issues as future work. Parity with Flux sounds like it will become important if/when we're really pushing migrations.

**Comment by gavincabbage** [comment]
Follow up work can be found in https://github.com/influxdata/idpe/issues/17634
